# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом.



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [ ]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://***/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [ ]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [ ]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [ ]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [ ]:

# Объявим новую функцию, которая подсчитает процентное соотношение
def percent(a):
    return '{:.1%} '.format(a.mean())


In [ ]:
# Выведем таблицу со сгрупированными строками по количеству детей(children) и столбцами
# количество клиентов(count), количество клиентов с задолжностью(sum) и нашей функцией(percent1),
# считающую процент должников от общего числа клиентов
# и отсортируем таблицу по процентному соотношению для удобства оценки данных
data.groupby('children')['debt'].agg(['count','sum', percent]).sort_values('percent')

count   sum percent
children                     
5             9     0   0.0% 
0         14091  1063   7.5% 
3           330    27   8.2% 
1          4808   444   9.2% 
2          2052   194   9.5% 
4            41     4   9.8%

**Вывод:** Реже всего имею задолжность бездетные, при наличии детей же их количество не особо влияет на вероятность иметь задолжность по кредит. Иначе говорят клиенты с детьми чаще могут иметь задолжности по кредитам, чем клиенты без детей.
Так же тяжело делать вывод о клиентах с 5 детьми, т.к. из 21525 клиентов, взявших кредит, всего 9 клиентов имеют 5 детей.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [ ]:
# Выведем таблицу со сгрупированными строками по семейному положению(family_status) и столбцами
# количество клиентов(count), количество клиентов с задолжностью(sum) и нашей функцией(percent1),
# считающую процент должников от общего числа клиентов
# и отсортируем таблицу по процентному соотношению для удобства оценки данных
data.groupby(['family_status'])['debt'].agg(['count','sum', percent]).sort_values('percent')

count  sum percent
family_status                            
вдовец / вдова           951   63   6.6% 
в разводе               1189   84   7.1% 
женат / замужем        12261  927   7.6% 
гражданский брак        4134  385   9.3% 
Не женат / не замужем   2796  273   9.8%

**Вывод:**
* Меньше всего задолжностей у овдовевших клиентов. Вероятно, это клиенты в возрасте, которые уже "нажили себе финансовую подушку".
* Больше всего задолжностей у клиентов, которые никогда не заключали брак.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [ ]:
# Выведем таблицу со сгрупированными строками по заранее созданным категориям согласно
# уровню дохода(total_income_category) и столбцами количество клиентов(count),
# количество клиентов с задолжностью(sum) и нашей функцией(percent1),
# считающую процент должников от общего числа клиентов
# и отсортируем таблицу по процентному соотношению для удобства оценки данных
data.groupby(['total_income_category'])['debt'].agg(['count','sum', percent]).sort_values('percent')

count   sum percent
total_income_category                     
D                        349    21   6.0% 
B                       5014   354   7.1% 
A                         25     2   8.0% 
C                      15921  1353   8.5% 
E                         22     2   9.1%

**Вывод:** Тяжело сделать вывод по данной зависимости, т.к. категории слишком сильно разнятся.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [ ]:
# Выведем таблицу со сгрупированными строками по заранее созданным категориям согласно
# целям кредитования(purpose_category) и столбцами количество клиентов(count),
# количество клиентов с задолжностью(sum) и нашей функцией(percent1),
# считающую процент должников от общего числа клиентов
# и отсортируем таблицу по процентному соотношению для удобства оценки данных
data.groupby(['purpose_category'])['debt'].agg(['count','sum', percent]).sort_values('percent')

count  sum percent
purpose_category                            
операции с недвижимостью  10751  780   7.3% 
проведение свадьбы         2313  183   7.9% 
операции с автомобилем     4279  400   9.3% 
получение образования      3988  369   9.3%

In [ ]:
# Можно вывести эти же данные методом .pivot_table()
data.pivot_table(index=['purpose_category'], values='debt', aggfunc=['count','sum',percent])

count  sum percent
                           debt debt    debt
purpose_category                            
операции с автомобилем     4279  400   9.3% 
операции с недвижимостью  10751  780   7.3% 
получение образования      3988  369   9.3% 
проведение свадьбы         2313  183   7.9%

**Вывод:**
* Меньше всего задолжностей у клиентов, кредитующихся с целью строительства/покупки/ремонта недвижимости.
* Больше всего задолжностей у клиентов, кредитующихся для операций с автомобилем и получения образования.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Вероятно пропуски были вызваны технологическими ошибками, т.к. получение кредита без информации об трудовом стаже и ежемесячного дохода, как правило, не возможно.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медианное значение менее чувствительно к резким перепадам в числах в большую и/или меньшую сторону, чем среднее значение. Если минимальное или максимальное значения имеют сильное отклонение от остальных значений, оно будет меньше влиять на результат в случае подсчета медианы. Если же сильного отклонения нет, медианное значение и так будет приближено к среднему.

### Шаг 4: общий вывод.

**ОБЩИЙ ВЫВОД**
Путём исследования надёжности заемщиков были обработаны полученные данные:
* испралены технологические ошибки(пропуски) и ошибки, вызванные человеческим фактором(ошибки в записи)
* заёмщики поделены по категориям согласно уровню зароботка и цели кредитования.

После обработки данных согласно поставленным вопросам сделаны следующие выводы:

*Есть ли зависимость между количеством детей и возвратом кредита в срок?*

* Да, зависимость есть:
    Реже всего имею задолжность бездетные, при наличии детей же их количество не особо влияет на вероятность иметь задолжность по кредит. Иначе говорят клиенты с детьми чаще могут иметь задолжности по кредитам, чем клиенты без детей.
* Но:
    Тяжело делать вывод о клиентах с 5 детьми, т.к. из 21525 клиентов, взявших кредит, всего 9 клиентов имеют 5 детей.

*Есть ли зависимость между семейным положением и возвратом кредита в срок?*

* Да, зависимость есть:
    * Меньше всего задолжностей у овдовевших клиентов. Вероятно, это клиенты в возрасте, которые уже "нажили себе финансовую подушку".
    * Больше всего задолжностей у клиентов, которые никогда не заключали брак.

*Есть ли зависимость между уровнем дохода и возвратом кредита в срок?*

* Нет, зависимости нет:
    Тяжело сделать вывод по данной зависимости, т.к. категории слишком сильно разнятся.

*Как разные цели кредита влияют на его возврат в срок?*

* Меньше всего задолжностей у клиентов, кредитующихся с целью строительства/покупки/ремонта недвижимости.
* Больше всего задолжностей у клиентов, кредитующихся для операций с автомобилем и получения образования.

**Рекомендации для компании**
Рекомендовал бы использовать рекламные рассылки с направлением на ипотечные кредиты, нацеленные на молодые семьи, не имеющих детей. Так же клиентам в более осознанном возрасте. Именно эти категории людей имеют меньше всего задолжностей по кредитам.